tao_compiler/mlir/disc/transforms/disc_transform_schedule.cc

```diff
@@ -0,0 +1,430 @@
+                                              ArrayRef<int64_t> tileSizes,
+                                              bool padded,
+                                              ArrayRef<int64_t> permutation) {
+  return b.create<transform_dialect::CacheReadOp>(
+      loc, target, anchor, tileLevels, tileSizes, padded, permutation);
+}
+
+transform_dialect::LowerMultiLevelPackToLoopOp buildLowerMultiLevelPackToLoop(
+    OpBuilder& b, Location& loc, Value target) {
+  return b.create<transform_dialect::LowerMultiLevelPackToLoopOp>(loc, target);
+}
+
+VectorizeOp buildVectorize(OpBuilder& b, Location& loc, Value target,
+                           bool vectorizePad) {
+  auto pdlType = pdl::OperationType::get(b.getContext());
+  return b.create<VectorizeOp>(loc, pdlType, target, vectorizePad);
+}
+
+transform_dialect::DISCBufferizeOp buildDISCBufferize(OpBuilder& b,
+                                                      Location& loc,
+                                                      Value target) {
+  auto pdlType = pdl::OperationType::get(b.getContext());
+  return b.create<transform_dialect::DISCBufferizeOp>(loc, pdlType, target);
+}
+
+void buildLowerVectors(OpBuilder& b, Location& loc, ArrayRef<int64_t> stages,
+                       StringRef contractionLowering,
+                       StringRef multireductionLowering,
+                       StringRef splitTransfers, bool unrollVectorTransfers,
+                       StringRef transposeLowering,
+                       bool transposeAvx2Lowering) {
+  b.create<transform_ext::LowerVectorsOp>(
+      loc, b.getI64ArrayAttr(stages), contractionLowering,
+      multireductionLowering, splitTransfers, unrollVectorTransfers,
+      transposeLowering, transposeAvx2Lowering);
+}
+
+LogicalResult aarch64GEMMDefaultScheduleFactory(PatternDescription& pd,
+                                                ModuleOp m) {
+  OpBuilder b(m);
+  b.setInsertionPointToStart(&m.getBodyRegion().front());
+  Location loc = m.getLoc();
+  MLIRContext* ctx = m->getContext();
+  auto seqOp = b.create<transform_ext::CanonicalizedSequenceOp>(
+      loc, TypeRange{}, transform::FailurePropagationMode::Propagate, Value{});
+  seqOp.getBody().push_back(new Block);
+  auto& bodyBlock = seqOp.getBody().front();
+  auto pdlOpType = pdl::OperationType::get(ctx);
+  bodyBlock.addArgument(pdl::OperationType::get(ctx), loc);
+  b.setInsertionPointToStart(&bodyBlock);
+  Value variant = bodyBlock.getArgument(0);
+
+  // %fill = transform.structured.match ops{["linalg.fill"]} in %variant
+  Value fill = buildMatchOp(b, loc, variant, {"linalg.fill"});
+  // %matmul = transform.structured.match ops{["linalg.matmul"]} in %arg1
+  Value matmul = buildMatchOp(b, loc, variant, {"linalg.matmul"});
+
+  // transform.structured.tile_to_foreach_thread_op %matmul num_threads [1, 1]
+  auto forEachThreadOp = buildTileToForEachThreadOp(b, loc, matmul, {1, 1});
```

Comment:
It seems to rewrite the schedule's pdl version in c++ ? Which kind do we prefer in our codebase?

Reply:
It's MLIR file with transform IR instead of PDL actually. The C++ version is used to provide a default schedule while we may override the default strategy for common fusion pattern (e.g. GEMM) by providing customized transform IR. Another use case is to provide codegen implementation for customized pattern. In that case, PDL may be useful to define the pattern.

it's worth noting that the C++ version is JIT generated. Thus it can be fine-tuned to make it suitable for specific shape setting. For example, we always do two level tiling in the current schedule. However, some tiled loops may only has 1 iteration if the shape of the GEMM is partial known (e.g., n, k of weight) and very small.  We have to take such information into consideration in order to make sure the transform IR is valid (e.g. some transform primitive depends on the number loop level).